enterprise/server/backends/distributed/distributed_test.go

```diff
@@ -0,0 +1,197 @@
+	flags.Set(t, "auth.enable_anonymous_usage", "true")
+	te := getTestEnv(t, emptyUserMap)
+	ctx, err := prefix.AttachUserPrefixToContext(context.Background(), te)
+	if err != nil {
+		t.Errorf("error attaching user prefix: %v", err)
+	}
+	return ctx
+}
+
+func waitUntilServerIsAlive(addr string) {
+	for {
+		_, err := net.DialTimeout("tcp", addr, 10*time.Millisecond)
+		if err == nil {
+			return
+		}
+	}
+}
+
+func waitUntilServerIsDead(addr string) {
+	for {
+		_, err := net.DialTimeout("tcp", addr, 10*time.Millisecond)
+		if err != nil {
+			return
+		}
+	}
+}
+
+func newDistributedCache(t *testing.T, te environment.Env, peer string, replicationFactor int, maxSizeBytes int64) *distributed.Cache {
+	mc, err := memory_cache.NewMemoryCache(maxSizeBytes)
+	if err != nil {
+		t.Fatal(err)
+	}
+	c, err := distributed.NewDistributedCache(te, mc, peer, heartbeatGroupName, replicationFactor)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return c
+}
+
+func TestDroppedNode(t *testing.T) {
+	te := getTestEnv(t, emptyUserMap)
+	ps := pubsub.NewTestPubSub()
+	te.SetPubSub(ps)
+	ctx := getAnonContext(t)
+
+	var liveNodes map[string]struct{}
+	hbc := heartbeat.NewHeartbeatChannel(te.GetPubSub(), "", heartbeatGroupName, func(nodes ...string) {
+		liveNodes = make(map[string]struct{}, 0)
+		for _, n := range nodes {
+			liveNodes[n] = struct{}{}
+		}
+	})
+	_ = hbc // keep hbc around to update liveNodes
+
+	tests := []struct {
+		replicas          int
+		replicationFactor int
+		replicasToFail    int
+	}{
+		{1, 1, 0},
+		{2, 2, 1},
+		{3, 2, 1},
+		{5, 3, 2},
+	}
+
+	for _, testStruct := range tests {
+		maxSizeBytes := int64(10000000) // 10MB
+		peers := make([]string, 0, testStruct.replicas)
+		caches := make(map[string]*distributed.Cache, testStruct.replicas)
+
+		log.Printf("TestStruct; %+v", testStruct)
+		for i := 0; i < testStruct.replicas; i++ {
+			peer := fmt.Sprintf("localhost:%d", app.FreePort(t))
+			peers = append(peers, peer)
+			caches[peer] = newDistributedCache(t, te, peer, testStruct.replicationFactor, maxSizeBytes)
+		}
+
+		// wait for the come up
+		for peer, _ := range caches {
+			waitUntilServerIsAlive(peer)
+		}
+
+		// wait until all nodes have advertised.
+		for {
+			if len(liveNodes) == testStruct.replicas {
+				break
+			}
+			time.Sleep(100 * time.Millisecond)
+		}
+
+		digests := make([]*repb.Digest, 0, 100)
+		for i := 0; i < 10; i += 1 {
+			d, buf := testdigest.NewRandomDigestBuf(t, 1500)
+			digests = append(digests, d)
+
+			randomPeer := peers[rand.Intn(len(peers))]
+			c := caches[randomPeer]
+			err := c.Set(ctx, d, buf)
+			if err != nil {
+				t.Fatalf("Error setting %q in cache: %s", d.GetHash(), err.Error())
+			}
+
+			randomPeer = peers[rand.Intn(len(peers))]
+			c = caches[randomPeer]
+			rbuf, err := c.Get(ctx, d)
+			if err != nil {
+				t.Fatalf("Error getting %q from cache: %s", d.GetHash(), err.Error())
+			}
+
+			// Compute a digest for the bytes returned.
+			d2, err := digest.Compute(bytes.NewReader(rbuf))
```

Comment:
should the test assert that this err is nil?

Reply:
Done.